### ISSUE TITLE
Authorize.Net MD5 end-of-life patch not working

### ISSUE THREAD

**Issue by treii28** [magento/magento2#21696]
Magento v2.2.4
Auth.net.md5.composer-2019-02-27-11-51-12.patch

Apply patch, remove Transkey and MD5 hash values, add new generated signature
run a payment through the cart using credit card

### Expected result (*)
Payment successful

### Actual result (*)
Initially I still had the Transkey and MD5 values inserted as there was nothing in the instructions from Magento/Authorize to remove them. It said something went wrong in the cart and checking the exception log showed:

"The transaction was declined because the response hash validation failed"

I contacted Authorize.net via a live chat and they suggested removing the Transkey and MD5 values as the original instructions didn't clearly state that. I removed those values and got the same error after clearing the cache and trying another transaction.
They suggested opening a support ticket. I did and got a canned response.

This morning I went through all the values and settings one by one, re-entering each to make sure the correct values were stored. (magento changes any secure credentials to 'dots') and re-saved. Ran another transaction through after clearing the cache and this time I get:

"Please enter a transaction ID to authorize this payment."

I note the release version of the code in the main branch still does not reflect the code changes, but Authorize.net is dropping MD5 support on the 28th of this month (originally the 14th).

**Comment by magento-engcom-team** [comment]
Hi @treii28. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@treii28 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by treii28** [comment]
Magento/authorize sent instructions to download the patch due to an end-of-life of md5 hash. There was a 1 week to live timeline so I began implementing the patch immediately.
The instructions were hard to follow as it told you to go to specific pages and follow specific links that did not exist. I eventually tracked the patch down the hard way and downloaded the version that said it was for v2.2.4 of magento and it didn't match the authorizenet module in v2.2.4 (Model/DirectPost.php was offset by -12 lines)
I managed to get the patch in place regardless and verified manually that all of the lines of code referenced in the patch file were in fact modified as instructed by the patch.

I created a transaction and got the errors listed above.

This is not a code change started by me, done by me, authorized by me, or otherwise done with my cooperation or consent. This is something I was told to do via an email from magento and it does not work. It causes the errors I listed in the original post.

I got the information on the patch from Magento. And it doesn't seem to work, thus I'm posting it here.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
HI @treii28 thank you for you report, i'm able to apply this patch by this instruction -> https://support.magento.com/hc/en-us/articles/360024368392 please try again maybe you was miss something.

**Comment by treii28** [comment]
I was able to apply the patch. That's not the problem. I worked around the bad instructions and got the patch to install and maintain using composer myself. The problem is THE PATCH DOESN'T WORK!

I'm still getting errors - or should say I'm now getting errors as the old method works fine on our production site, granted until they permanently change it. But the new code won't submit a successful payment.


**Actual result**
Initially I still had the Transkey and MD5 values inserted as there was nothing in the instructions from Magento/Authorize to remove them. It said something went wrong in the cart and checking the exception log showed:

_"The transaction was declined because the response hash validation failed"_

I contacted Authorize.net via a live chat and they suggested removing the Transkey and MD5 values as the original instructions didn't clearly state that. I removed those values and got the same error after clearing the cache and trying another transaction.
They suggested opening a support ticket. I did and got a canned response.

This morning I went through all the values and settings one by one, re-entering each to make sure the correct values were stored. (magento changes any secure credentials to 'dots') and re-saved. Ran another transaction through after clearing the cache and this time I get:

_"Please enter a transaction ID to authorize this payment."_



(I also do not seem to be the only one having this problem as I have found at least half-a-dozen similar threads on various community boards reporting the same or similar issues after applying the official patch)

**Comment by muramasatheninja** [comment]
Having the same problem. Following in hopes of a fix.

**Comment by treii28** [comment]
I appears they have now extended the deadline yet again, so presumably a lot of folks are having problems. But meanwhile, I can't seem to get anyone involved with the magento plugin to even acknowledge there is any problem.

**Comment by makvital** [comment]
I do have same problem. Actually there's "transaction approval from Authorize.Net". As per the Debug log it's all perfect both request and response. 

But still in Magento it displays the below error.
_"The transaction was declined because the response hash validation failed"_

**Comment by ghost** [comment]
@treii28 now i'm recheck this.

**Comment by ghost** [comment]
@treii28 double checked and all works fine, did you enter new key from admin ?

**Comment by makvital** [comment]
@engcom-backlog-nazar here's the steps I followed. 

- Removed  "/a" and "/b" from the patch in the "Auth.net.md5.composer-2019-02-27-11-51-12.patch".

- Applied the patch "Auth.net.md5.composer-2019-02-27-11-51-12.patch" using the following command "patch -p0 > Auth.net.md5.composer-2019-02-27-11-51-12.patch". 

- Cleared the cache

- Fetch the Signature Key from Authorize.Net account from "Settings >> API Credentials & Keys"

- Added the Signature Key in Authorize.Net Direct Post Payment Method Configuration in the Magento. Emptied "MD5 Hash". 

- Ran a test transaction, found it was approved in Authroize.Net dashboard(Received the Payment confirmation email from Authorize.Net). Debug log too shows the transaction was approved.  

- But an error occurred in Checkout, "The transaction was declined because the response hash validation failed".

Anything not followed here or wrong?

**Comment by treii28** [comment]
Yup, same results here as makvital

The item also disapeers from the cart, but the checkout process doesn't complete.

**Comment by treii28** [comment]
There is also a similar discussion on the magento support blog: [https://community.magento.com/t5/Magento-DevBlog/Authorize-Net-Direct-Post-Patch-for-M2-and-M1/bc-p/124989#M445](https://community.magento.com/t5/Magento-DevBlog/Authorize-Net-Direct-Post-Patch-for-M2-and-M1/bc-p/124989#M445)

**Comment by treii28** [comment]
Why was this ticket closed? The initial issue was never address, it is still producing an error for more than one person involved in this thread alone. (as well as others on other threads linked in the comments above)

**Comment by ghost** [comment]
@treii28 yes, i'm found internal ticket in jira and, this bug fixed in 2.3-develop and will be available on 2.3.1 release,we  will wait for backport and patch https://github.com/magento/magento2ce/pull/3889

**Comment by treii28** [comment]
I get a 404 on that link

**Comment by ghost** [comment]
@treii28 yes this is private repository.

**Comment by vipin12117** [comment]
@magento team: any update on error ?

**Comment by treii28** [comment]
How long until there is an updated patch?

**Comment by vipin12117** [comment]
Magento team and Authnet team both not accepting their mistakes. Authnet api does not give response with SHA key which needs to verify.

**Comment by Ctucker9233** [comment]
@engcom-backlog-nazar that PR link goes to a 404 page. Can you fix the link and i'll attempt to backport.

**Comment by vipin12117** [comment]
Currently we solved this problem by returned true in isValidHash function.

**Comment by ghost** [comment]
@Ctucker9233 this link to private repository, this already fixed in 2.2-dev branch. thanks for collaboration :+1:

**Comment by magecomp** [comment]
Fixed after patching. 
Popup error was due to site was on maintainance mode.
Removing maintainance mode helped to get rid of that annoying popup at checkout. :)

**Comment by makvital** [comment]
Can you provide the link for updated patch? I couldn't find it in the resource. https://magento.com/tech-resources/download#download2280

**Comment by magecomp** [comment]
Try this
https://github.com/magecomp/Magento-Pre-Patched-Files/tree/master/Authorize.net%20patch%20files

**Comment by Ctucker9233** [comment]
@engcom-backlog-nazar can you link the 2.2 PR then? Thanks.

**Comment by irajneeshgupta** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta, here is your [Magento instance](https://i-21696-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-21696-2-2-develop.instances.magento-community.engineering/admin](https://i-21696-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.
